src/IO/ParallelReadBuffer.cpp

```diff
@@ -0,0 +1,288 @@
+
+    while (!read_workers.empty() && (offset < current_position || !offset_is_in_range(read_workers.front()->range)))
+    {
+        read_workers.pop_front();
+    }
+
+    if (!read_workers.empty())
+    {
+        auto & front_worker = read_workers.front();
+        auto & segments = front_worker->segments;
+        current_position = front_worker->range.from;
+        while (true)
+        {
+            next_condvar.wait(lock, [&] { return emergency_stop || !segments.empty(); });
+
+            if (emergency_stop)
+                handleEmergencyStop();
+
+            auto next_segment = front_worker->nextSegment();
+            if (static_cast<size_t>(offset) < current_position + next_segment.size())
+            {
+                current_segment = std::move(next_segment);
+                working_buffer = internal_buffer = Buffer(current_segment.data(), current_segment.data() + current_segment.size());
+                current_position += current_segment.size();
+                pos = working_buffer.end() - (current_position - offset);
+                addReaders(lock);
+                return offset;
+            }
+
+            current_position += next_segment.size();
+        }
+    }
+
+    lock.unlock();
+    finishAndWait();
+
+    reader_factory->seek(offset, whence);
+    all_completed = false;
+    read_workers.clear();
+
+    current_position = offset;
+    resetWorkingBuffer();
+
+    emergency_stop = false;
+
+    lock.lock();
+    addReaders(lock);
+    return offset;
+}
+
+std::optional<size_t> ParallelReadBuffer::getTotalSize()
+{
+    std::lock_guard lock{mutex};
+    return reader_factory->getTotalSize();
+}
+
+off_t ParallelReadBuffer::getPosition()
+{
+    return current_position - available();
+}
+
+bool ParallelReadBuffer::currentWorkerReady() const
+{
+    return !read_workers.empty() && (read_workers.front()->finished || !read_workers.front()->segments.empty());
+}
+
+bool ParallelReadBuffer::currentWorkerCompleted() const
+{
+    return !read_workers.empty() && read_workers.front()->finished && read_workers.front()->segments.empty();
+}
+
+void ParallelReadBuffer::handleEmergencyStop()
+{
+    if (background_exception)
+        std::rethrow_exception(background_exception);
+    else
+        throw DB::Exception(ErrorCodes::LOGICAL_ERROR, "Emergency stop");
+}
+
+bool ParallelReadBuffer::nextImpl()
+{
+    if (all_completed)
+        return false;
+
+    while (true)
+    {
+        std::unique_lock lock(mutex);
+        next_condvar.wait(
+            lock,
+            [this]()
+            {
+                /// Check if no more readers left or current reader can be processed
+                return emergency_stop || currentWorkerReady();
+            });
+
+        if (emergency_stop)
+            handleEmergencyStop();
+
+        bool worker_removed = false;
+        /// Remove completed units
+        while (!read_workers.empty() && currentWorkerCompleted())
```

Comment:
Should we also check in this loop `&& !emergency_stop`?

Reply:
Technically we can check for `emergency_stop` every few steps so I'm not sure if it makes sense adding here as opposed to somewhere else in the loop. 
What I can do is add the check here and move the `if (emergency_stop) handleEmergencyStop();` bellow it.